angr/knowledge_plugins/labels.py

```diff
@@ -1,66 +1,237 @@
+from collections import defaultdict, Iterable
+
+from bidict import bidict
+
 from .plugin import KnowledgeBasePlugin
+from ..misc.ux import deprecated
+
+
+class LabelsPlugin(KnowledgeBasePlugin):
+    """Storage for program labels. Access with kb.labels.    
+    
+    A label "name" is a label named "name", duh. Labels can be grouped into the seperate namespaces.
+    An address is allowed to have a different labels within different namespace, but it is not allowed
+    for address to have a different labels within a single namespace.
+    
+    The global namespace is the empty string.
+       
+    :var _global_ns:    The name of the global namespace.
+    :type _global_ns:   str
+    """
+
+    _global_ns = ''
+
+    def __init__(self, kb, copy=False):
+        super(KnowledgeBasePlugin, self).__init__()
+        self._namespaces = defaultdict(bidict)
 
+        if not copy:
+            # TODO: This should be done somewhere else. Here for compat reasons only.
+            for obj in kb._project.loader.all_objects:
+                for k, v in obj.symbols_by_addr.iteritems():
+                    if v.name:
+                        # Original logic implies overwriting existing labels.
+                        if self.get_addr(v.name):
+                            self.del_label(v.name)
+                        self.set_label(v.rebased_addr, v.name)
+                try:
+                    for k, v in obj.plt.iteritems():
+                        # Original logic implies overwriting existing labels.
```

Comment:
I definitely didn't intend there to be overwriting semantics in the initialization code...

The original logic in the label-_setting_ code was designed to mirror IDA as closely as possible, where "naming" a spot would override the existing label, despite the ability to have multiple names for an address, since it would be silly to rename an address several times and have all those names still be present. Since we're working with an API and not a user interface, you can pick whatever paradigm you want.